From e555c1cccc79067f1688c8b3795b2740654af526 Mon Sep 17 00:00:00 2001
From: Piyush Mehta <piyush.mehta@xilinx.com>
Date: Tue, 17 Aug 2021 07:26:45 +0000
Subject: [PATCH 2/4] usb: dwc3: core: add transceiver delay for required
 UTMI/ULPI phy

GUSB2PHYCFG.XCVRDLY adds transceiver delay from the time when the
transceiver select is set to 2'b00 (HS) to the time the TxValid is
driven to 0 for sending the chirp-K.

When transceiver_delay_quirk property set then dwc3 core introduce 2.5us
delay. This delay is required for some UTMI/ULPI PHYs.

Adding this option we can enable/disable XCVRDLY bit for transceiver delay.

Signed-off-by: Piyush Mehta <piyush.mehta@xilinx.com>
---
 drivers/usb/dwc3/core.c | 12 ++++++++++++
 drivers/usb/dwc3/core.h |  4 ++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index a8996a651279..40f3ceb4f8cc 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -852,6 +852,16 @@ static int dwc3_phy_setup(struct dwc3 *dwc)
 	if (dwc->ulpi_ext_vbus_drv)
 		reg |= DWC3_GUSB2PHYCFG_ULPIEXTVBUSDRV;
 
+	/*
+	 * Enables a delay between the assertion of the UTMI/ULPI Transceiver
+	 * Select signal (for HS) and the assertion of the TxValid signal
+	 * during a HS Chirp. When this bit is set to 1, a delay of 2.5us.
+	 */
+	if (dwc->transceiver_delay_quirk)
+		reg |= DWC3_GUSB2PHYCFG_XCVRDLY;
+	else
+		reg &= ~DWC3_GUSB2PHYCFG_XCVRDLY;
+
 	dwc3_writel(dwc->regs, DWC3_GUSB2PHYCFG(0), reg);
 
 	return 0;
@@ -1641,6 +1651,8 @@ static void dwc3_get_properties(struct dwc3 *dwc)
 				"snps,dis_u2_susphy_quirk");
 	dwc->dis_enblslpm_quirk = device_property_read_bool(dev,
 				"snps,dis_enblslpm_quirk");
+	dwc->transceiver_delay_quirk = device_property_read_bool(dev,
+				"snps,transceiver_delay_quirk");
 	dwc->dis_u1_entry_quirk = device_property_read_bool(dev,
 				"snps,dis-u1-entry-quirk");
 	dwc->dis_u2_entry_quirk = device_property_read_bool(dev,
diff --git a/drivers/usb/dwc3/core.h b/drivers/usb/dwc3/core.h
index f2d42c98aaed..7486f5a3d021 100644
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -296,6 +296,7 @@
 #define DWC3_GUSB2PHYCFG_SUSPHY		BIT(6)
 #define DWC3_GUSB2PHYCFG_ULPI_UTMI	BIT(4)
 #define DWC3_GUSB2PHYCFG_ENBLSLPM	BIT(8)
+#define DWC3_GUSB2PHYCFG_XCVRDLY	BIT(9)
 #define DWC3_GUSB2PHYCFG_PHYIF(n)	(n << 3)
 #define DWC3_GUSB2PHYCFG_PHYIF_MASK	DWC3_GUSB2PHYCFG_PHYIF(1)
 #define DWC3_GUSB2PHYCFG_USBTRDTIM(n)	(n << 10)
@@ -1107,6 +1108,8 @@ struct dwc3_scratchpad_array {
  * @dis_u2_susphy_quirk: set if we disable usb2 suspend phy
  * @dis_enblslpm_quirk: set if we clear enblslpm in GUSB2PHYCFG,
  *                      disabling the suspend signal to the PHY.
+ * @transceiver_delay_quirk: set if we enable to add delay in GUSB2PHYCFG,
+ *                      disabling the remove  transceiver delay to the PHY.
  * @dis_u1_entry_quirk: set if link entering into U1 state needs to be disabled.
  * @dis_u2_entry_quirk: set if link entering into U2 state needs to be disabled.
  * @dis_rxdet_inp3_quirk: set if we disable Rx.Detect in P3
@@ -1344,6 +1347,7 @@ struct dwc3 {
 	unsigned		dis_u3_susphy_quirk:1;
 	unsigned		dis_u2_susphy_quirk:1;
 	unsigned		dis_enblslpm_quirk:1;
+	unsigned		transceiver_delay_quirk:1;
 	unsigned		dis_u1_entry_quirk:1;
 	unsigned		dis_u2_entry_quirk:1;
 	unsigned		dis_rxdet_inp3_quirk:1;
-- 
2.17.1

